### PR TITLE
fix: use _comp_compgen for `COMPREPLY=($(compgen ...))`

### PR DESCRIPTION
**Pull request by akinomyoga**
These are relatively trivial ones for the `_comp_compgen` rewriting.

~~**edit**: #952 is blocking more, so could you see #952 before this?~~ Merged

Note: The commits "fix: use _comp_compgen for COMPREPLY=(...)" are essentially generated by `sed`:

```
sed -E "s/$path/$rep/g"

with

pat='(COMPREPLY)=\(\$\(compgen ([^()]*) -- "\$(cur|\{cur-?\})"[[:space:]]*\)\)'   rep='_comp_compgen -- \2'
pat='(COMPREPLY)\+=\(\$\(compgen ([^()]*) -- "\$(cur|\{cur-?\})"[[:space:]]*\)\)' rep='_comp_compgen -a -- \2'
pat='(COMPREPLY)=\(\$\(compgen ([^()]*) -- ("[^"()]+")[[:space:]]*\)\)'           rep='_comp_compgen -c \3 -- \2'
pat='(COMPREPLY)\+=\(\$\(compgen ([^()]*) -- ("[^"()]+")[[:space:]]*\)\)'         rep='_comp_compgen -ac \3 -- \2'
pat='(COMPREPLY)=\(\$\(compgen ([^()]*)\)\)'                                      rep='_comp_compgen -R -- \2'
pat='(COMPREPLY)\+=\(\$\(compgen ([^()]*)\)\)'                                    rep='_comp_compgen -aR -- \2'

```

### CODE REVIEW

**Comment by akinomyoga** [comment]
- d660f8a60de8eb5873203a1d53e3774e7cf475ab In rebasing, I already found four oversights of `-a` that should not have been removed...

**Comment by akinomyoga** [comment]
The cases in f00dfd0b30256a4173e3f8370dda00af6e3d7e12 and ac886949cfc2aaa3f18f8aca56f8bd26b48f012d are the ones originally without specifying any `-- "..."`. I suspect some of them were unintentionally forgotten to add `-- "$cur"`.
- The completion generation for single characters such as `-W '{0..9}'` is fine without `-- "$cur"` because any filtering using `"$cur"` would generate all or nothing.
- The generated completions containing `$cur` as a prefix would also be fine (feh, tar, wget, rcs), though some of them might need proper quoting against the -W splitting.
- `xgamma` appears to try to remove the beginning `:`.
- suspect `hping2` (`1[TAB]` would not complete)
- sus `lintian`, `useradd`, `usermod`

**Review by scop — COMMENTED** [review]
LGTM! Pre-approved with some comments addressed.

File: completions/tune2fs [hunk 1]
```diff
@@ -10,13 +10,13 @@ _comp_cmd_tune2fs()
...
         -*e)
-            COMPREPLY=($(compgen -W 'continue remount-ro panic' -- "$cur"))
+            _comp_compgen -- -W 'continue remount-ro panic'
             return
             ;;
         -*g)
             _gids
             ((${#COMPREPLY[@]})) &&
-                COMPREPLY=($(compgen -g -W '"${COMPREPLY[@]}"' -- "$cur"))
+                _comp_compgen -- -g -W '"${COMPREPLY[@]}"'
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tune2fs:19]: Ah, this is my mistake in [`76b7925`](https://github.com/scop/bash-completion/commit/76b7925a5f31f0c24ea3118b0fc4d5a208fa8ece#diff-ae91d3cf5b9a0757b0f0aacdd668090678d1139ee65cc9422081fa9a349283e1) two years before. I dismissed the fact that `compgen -g/-u -W '"${COMPREPLY[@]}"'` would generate group/user ids even when `COMPREPLY` is empty. I checked the entire commit 76b7925a5f31f0c24ea3118b0fc4d5a208fa8ece again, but the mistake was only found in `tune2fs`.

The original code before the breaking commit appears to have tried to filter the result of `_gids` (the same as in `usermod`/`useradd`), which is actually unneeded, so it should simply be `_comp_compgen -a -- -g` as you suggest.

[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tune2fs]: 4acc4d5fd9873174706145589fe17642b906145e

File: completions/useradd [hunk 2]
```diff
@@ -30,9 +30,9 @@ _comp_cmd_useradd()
             ;;
         --gid | -${noargopts}g)
             _gids
-            COMPREPLY+=($(compgen -g))
+            _comp_compgen -aR -- -g
             ((${#COMPREPLY[@]})) &&
-                COMPREPLY=($(compgen -W '"${COMPREPLY[@]}"' -- "$cur"))
+                _comp_compgen -- -W '"${COMPREPLY[@]}"'
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/useradd]: 4acc4d5fd9873174706145589fe17642b906145e

There were also similar cases in `ifup`, `ip`, and `smartctl`, so I also included the fixes to these.

File: completions/usermod [hunk 3]
```diff
@@ -26,9 +26,9 @@ _comp_cmd_usermod()
             ;;
         --gid | -${noargopts}g)
             _gids
-            COMPREPLY+=($(compgen -g))
+            _comp_compgen -aR -- -g
             ((${#COMPREPLY[@]})) &&
-                COMPREPLY=($(compgen -W '"${COMPREPLY[@]}"' -- "$cur"))
+                _comp_compgen -- -W '"${COMPREPLY[@]}"'
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/usermod]: 4acc4d5fd9873174706145589fe17642b906145e

File: completions/useradd (hunk 2 above)
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/useradd]: Also, before applying 4acc4d5fd9873174706145589fe17642b906145e, I changed the behavior of  `_comp_cmd_smartctl__device` (c447a81f677fef7ec23a576299b15f8666259627) so it always replaces `COMPREPLY`. Previously it replaces or appends the new elements depending on `cur`, which doesn't seem to be consistent. The caller is in the context where `COMPREPLY` is empty.

File: completions/gnome-mplayer [hunk 4]
```diff
@@ -10,15 +10,15 @@ _comp_cmd_gnome_mplayer()
             return
             ;;
         --showcontrols | --showsubtitles | --autostart)
-            COMPREPLY=($(compgen -w '0 1' -- "$cur"))
+            _comp_compgen -- -w '0 1'
```
[the remark this answers is not in the document]
Comment (scop) [reply completions/gnome-mplayer:13]: Maybe this comment went unnoticed/accidentally resolved without change?

File: completions/hping2 [hunk 5]
```diff
@@ -17,7 +17,7 @@ _comp_cmd_hping2()
             return
             ;;
         --tos | -${noargopts}o)
-            COMPREPLY=($(compgen -W '02 04 08 10'))
+            _comp_compgen -R -- -W '02 04 08 10'
```
[the remark this answers is not in the document]
Comment (scop) [reply completions/hping2:20]: Maybe this comment went unnoticed/accidentally resolved without change?

File: completions/gnome-mplayer (hunk 4 above)
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/gnome-mplayer:13]: Oh, I think I mistakenly closed the conversations and forgot about them. I'll apply them.

File: completions/hping2 (hunk 5 above)
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/hping2]: Applied in 3736999bc5f7868e65029410f304ecbad658347f

File: completions/gnome-mplayer (hunk 4 above)
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/gnome-mplayer]: Applied in c33777b3b743a52356b690b479de12b3571d7a24

Hmm, it is strange. The text box for inputting the commit message for GitHub UI [Commit Suggested Changes] has memorized the commit message that I input the previous time. Maybe I forgot to press the commit button, or the transactions have failed halfway?

[the remark this answers is not in the document]
Comment (scop) [reply completions/gnome-mplayer]: I've had that happen once or twice as well, have thought that those had been an oversight on my part. But who knows :shrug:

**Review by scop — APPROVED** [review]

**Pull request merged**